keras/layers/convolutional_recurrent.py

```diff
@@ -1,172 +1,585 @@
+from ..utils.generic_utils import has_arg
 
 
-class ConvRecurrent2D(Recurrent):
-    """Abstract base class for convolutional recurrent layers.
-
-    Do not use in a model -- it's not a functional layer!
+class ConvRNN2D(Layer):
+    """Base class for recurrent layers.
 
     # Arguments
-        filters: Integer, the dimensionality of the output space
-            (i.e. the number output of filters in the convolution).
-        kernel_size: An integer or tuple/list of n integers, specifying the
-            dimensions of the convolution window.
-        strides: An integer or tuple/list of n integers,
-            specifying the strides of the convolution.
-            Specifying any stride value != 1 is incompatible with specifying
-            any `dilation_rate` value != 1.
-        padding: One of `"valid"` or `"same"` (case-insensitive).
-        data_format: A string,
-            one of `channels_last` (default) or `channels_first`.
-            The ordering of the dimensions in the inputs.
-            `channels_last` corresponds to inputs with shape
-            `(batch, time, ..., channels)`
-            while `channels_first` corresponds to
-            inputs with shape `(batch, time, channels, ...)`.
-            It defaults to the `image_data_format` value found in your
-            Keras config file at `~/.keras/keras.json`.
-            If you never set it, then it will be "channels_last".
-        dilation_rate: An integer or tuple/list of n integers, specifying
-            the dilation rate to use for dilated convolution.
-            Currently, specifying any `dilation_rate` value != 1 is
-            incompatible with specifying any `strides` value != 1.
-        return_sequences: Boolean. Whether to return the last output
+        cell: A RNN cell instance. A RNN cell is a class that has:
+            - a `call(input_at_t, states_at_t)` method, returning
+                `(output_at_t, states_at_t_plus_1)`. The call method of the
+                cell can also take the optional argument `constants`, see
+                section "Note on passing external constants" below.
+            - a `state_size` attribute. This can be a single integer
+                (single state) in which case it is
+                the number of channels of the recurrent state
+                (which should be the same as the number of channels of the cell output).
+                This can also be a list/tuple of integers
+                (one size per state). In this case, the first entry
+                (`state_size[0]`) should be the same as
+                the size of the cell output.
+            It is at the moment not possible to stack ConvLSTM2DCells because
+            the stack class isn't implemented yet. It is also not recommended to
+            use ConvRNN2D with other cell types than ConvLSTM2DCell because
+            some things are assumed to be true to simplify the implementation.
```

Comment:
Implementation-level comments should be left as code comments, and not included in the user-facing docs.